Rename `uniform` to `parameter` across the engine

Chaosus opened the pull request:
Changes:

- `ShaderMaterial::set_shader_uniform/get_shader_uniform` renamed to `set_shader_param/get_shader_param`.
- All other visual shader classes `<Type>Uniform` renamed to `<Type>Parameter`.
- Other uniform methods, parameter and variables renamed to parameters in the `VisualShaderEditor`.
- Separates `VisualShaderNodeTextureUniform` to `VisualShaderNodeTextureParameter` and `VisualShaderNodeTexture2DParameter`.
- Changed `default_value` type in `VisualShaderNodeVec4Parameter` from `Quaternion` to `Vector4`.
- Added compatibility classes and changed docs accordingly.

Target to close https://github.com/godotengine/godot-proposals/issues/5123 hope I did everything correct (as far as I remember the last bikeshedding meeting)

Chaosus committed:
Rename `uniform` to `parameter` across the engine

QbieShay commented:
Hey @Chaosus, thanks a lot for taking the time to do this!
I have two questions:
1. Is the entry for "global shader unifrom" in project setting also changed?
2. I am not sure if they preferred `set_shader_param` or `set_shader_parameter`? (@akien-mga ?)
The rest looks good to me! (Note I didn't test, i can't this weekend)

Chaosus commented:
```
Is the entry for "global shader unifrom" in project setting also changed?
```
In the project setting, it's called `Shader Globals` - how do you want to change it?

![image](https://user-images.githubusercontent.com/3036176/187065660-4f1c1162-0194-49d4-a11b-0672acf59c5c.png)

Also, I didn't change the name of the methods in the `RenderingServer` and `MaterialStorage` - I'm not sure that it is a good thing to do, and @reduz would certainly against it.

clayjohn commented:
> Also, I didn't change the name of the methods in the RenderingServer and MaterialStorage - I'm not sure that it is a good thing to do, and @reduz would certainly against it.

I think anything exposed to users should use "parameter". I don't think it is a big deal to change the naming in RenderingServer and MaterialStorage. After all, reduz originally named them with parameter and it was just changed to "uniform" recently (in https://github.com/godotengine/godot/pull/59840 and https://github.com/godotengine/godot/pull/59844)

Chaosus commented:
Okay, if this is desired I will make a patch to do it ^^

Zireael07 commented:
> I think anything exposed to users should use "parameter". 

Yep, should be consistent whichever way the cookie crumbles

Chaosus commented:
@clayjohn Alright, I've renamed uniform term to parameter across all functions and enumeration members:

![image](https://user-images.githubusercontent.com/3036176/187150232-4350fbc0-7a86-48fc-a4ac-c393226df4db.png)

nathanfranke commented:
If anyone disagrees with the naming in https://github.com/godotengine/godot/pull/64092, let me know. I decided `shader_param/` for compatibility/simplicity, and since the internal properties are not user facing (users should really be using `set_shader_param("foo", bar)`)

akien-mga commented:
> If anyone disagrees with the naming in #64092, let me know. I decided `shader_param/` for compatibility/simplicity, and since the internal properties are not user facing (users should really be using `set_shader_param("foo", bar)`)

We should fully spell out `parameter` everywhere (it's already `set_shader_parameter` in this PR).

It would be useful to have compatibility code for `shader_param` (before #59844) and `shader_uniform` (before this PR) to ease transition from 3.x and earlier alphas.

Chaosus commented:
I've added `set_shader_param` -> `set_shader_parameter` to the project converter. Also renamed bunch of parameter names and other internal functions. I leave only some internal classes untouched such as `RD::Uniform`, `UniformSet` etc - do I need to rename them too?

clayjohn commented:
> I've added `set_shader_param` -> `set_shader_parameter` to the project converter. Also renamed bunch of parameter names and other internal functions. I leave only some internal classes untouched such as `RD::Uniform`, `UniformSet` etc - do I need to rename them too?

No, RD::Uniform and Uniform set etc. should all remain as is. Just the user-facing functions to set a single uniform should be changed. I'll review now!

clayjohn commented on drivers/vulkan/rendering_device_vulkan.h:
```diff
@@ -546,7 +546,7 @@ class RenderingDeviceVulkan : public RenderingDevice {
 
 	struct UniformInfo {
-		UniformType type = UniformType::UNIFORM_TYPE_MAX;
+		ParameterType type = ParameterType::PARAMETER_TYPE_MAX;
```
I don't think we should change the RenderingDevice API. Its better to keep this closer to the actual names. It won't be used by beginners either, i.e. you have to know what a uniform is in order to use the RD API

clayjohn commented on scene/resources/shader.cpp:
```diff
@@ -107,7 +107,7 @@ void Shader::get_shader_uniform_list(List<PropertyInfo> *p_params, bool p_get_gr
 	_update_shader();
 
 	List<PropertyInfo> local;
```
This function should be renamed to ``get_shader_parameter_list``

clayjohn commented on scene/resources/shader.cpp:
```diff
@@ -176,8 +176,8 @@ bool Shader::is_text_shader() const {
-bool Shader::has_uniform(const StringName &p_param) const {
-	return params_cache.has("shader_uniform/" + p_param);
+bool Shader::has_parameter(const StringName &p_name) const {
+	return params_cache.has("shader_uniform/" + p_name);
```
Probably should be ``shader_parameter``, but I wonder where the variables are set in the cache

clayjohn commented on scene/resources/material.cpp:
```diff
@@ -384,15 +384,15 @@ void ShaderMaterial::_shader_changed() {
+	if ((f == "get_shader_parameter" || f == "set_shader_parameter") && p_idx == 0) {
 		if (shader.is_valid()) {
 			List<PropertyInfo> pl;
 			shader->get_shader_uniform_list(&pl);
```
In this function "shader_uniform" should be "shader_parameter"

clayjohn replied:
Also backwards compatibility should be added to ``_set()`` and ``_get()``

clayjohn commented on servers/register_server_types.cpp:
```diff
@@ -216,7 +216,7 @@ void register_server_types() {
 	GDREGISTER_CLASS(RDSamplerState);
 	GDREGISTER_CLASS(RDVertexAttribute);
-	GDREGISTER_CLASS(RDUniform);
+	GDREGISTER_CLASS(RDParameter);
```
Should be reverted

clayjohn commented on servers/rendering/rendering_device.h:
```diff
@@ -42,7 +42,7 @@ class RDSamplerState;
 class RDVertexAttribute;
 class RDShaderSource;
 class RDShaderSPIRV;
-class RDUniforms;
```
All of this should be reverted. RenderingDevice should continue using "uniform"

clayjohn reviewed (commented):
Looks mostly good. Great work! I think everything in RenderingDevice should continue to use "uniform". So you'll have to undo those changes unfortunately. I think it is best it maintains the original name as people using it will be more familiar with uniforms as uniform has a specific meaning in the context of graphics APIs that is incompatible with the more general term "parameter". Parameter is fine for the Godot-shader API though, so it is suitable for user-facing code in the editer, scene nodes and RenderingServer

Chaosus commented:
@clayjohn Done I think - check again

reduz commented:
The actual problem is that uniform term is unfriendly to VisualShader for artists, everywhere else it should be uniform, which is the right term. "parameter" is not more friendly for programmers, its more confusing because anything can be a parameter.

reduz commented on drivers/vulkan/rendering_device_vulkan.h:
```diff
@@ -1120,9 +1120,9 @@ class RenderingDeviceVulkan : public RenderingDevice {
-	virtual RID uniform_set_create(const Vector<Uniform> &p_uniforms, RID p_shader, uint32_t p_shader_set);
-	virtual bool uniform_set_is_valid(RID p_uniform_set);
-	virtual void uniform_set_set_invalidation_callback(RID p_uniform_set, InvalidationCallback p_callback, void *p_userdata);
+	virtual RID uniform_set_create(const Vector<Uniform> &p_parameters, RID p_shader, uint32_t p_shader_set);
```
RenderingDevice needs to be left alone, these are fine.

reduz commented on servers/rendering/renderer_geometry_instance.h:
```diff
@@ -59,7 +59,7 @@ class RenderGeometryInstance {
 	virtual void set_use_lightmap(RID p_lightmap_instance, const Rect2 &p_lightmap_uv_scale, int p_lightmap_slice_index) = 0;
 	virtual void set_lightmap_capture(const Color *p_sh9) = 0;
-	virtual void set_instance_shader_uniforms_offset(int32_t p_offset) = 0;
+	virtual void set_instance_shader_parameters_offset(int32_t p_offset) = 0;
```
This is OK as uniforms

reduz commented on servers/rendering/renderer_geometry_instance.h:
```diff
@@ -140,7 +140,7 @@ class RenderGeometryInstanceBase : public RenderGeometryInstance {
 	virtual void set_use_baked_light(bool p_enable) override;
 	virtual void set_use_dynamic_gi(bool p_enable) override;
-	virtual void set_instance_shader_uniforms_offset(int32_t p_offset) override;
+	virtual void set_instance_shader_parameters_offset(int32_t p_offset) override;
```
Also OK as uniforms

reduz commented on servers/rendering/renderer_geometry_instance.cpp:
```diff
@@ -117,7 +117,7 @@ void RenderGeometryInstanceBase::set_use_dynamic_gi(bool p_enable) {
 }
 
-void RenderGeometryInstanceBase::set_instance_shader_uniforms_offset(int32_t p_offset) {
+void RenderGeometryInstanceBase::set_instance_shader_parameters_offset(int32_t p_offset) {
```
This should be ok as uniforms

reduz commented on servers/rendering_server.h:
```diff
@@ -1469,20 +1469,20 @@ class RenderingServer : public Object {
+	virtual Variant global_shader_parameter_get(const StringName &p_name) const = 0;
+	virtual GlobalShaderParameterType global_shader_parameter_get_type(const StringName &p_name) const = 0;
 
 	virtual void global_shader_uniforms_load_settings(bool p_load_textures) = 0;
```
Missed these

reduz reviewed (changes requested):
It also marks vulkan_core.h as changed, it should not be?

Chaosus commented:
```It also marks vulkan_core.h as changed, it should not be?```

Sorry, due to formatting suddenly, reverted it back

reduz commented on drivers/gles3/rasterizer_scene_gles3.cpp:
```diff
@@ -1736,7 +1736,7 @@ void RasterizerSceneGLES3::render_scene(RID p_render_buffers, const CameraData *
 	//////////
 
-	GLuint global_buffer = GLES3::MaterialStorage::get_singleton()->global_shader_uniforms_get_uniform_buffer();
+	GLuint global_buffer = GLES3::MaterialStorage::get_singleton()->global_shader_parameters_get_parameter_buffer();
```
it should be:
`global_shader_parameters_get_uniform_buffer`

akien-mga commented:
Thanks!
